x/farming/keeper/reward.go

```diff
@@ -159,3 +147,117 @@ func (k Keeper) Harvest(ctx sdk.Context, farmerAcc sdk.AccAddress, stakingCoinDe
+		}
+	}
+
+	for _, plan := range plans {
+		farmingPoolAcc := plan.GetFarmingPoolAddress()
+		farmingPool := farmingPoolAcc.String()
+
+		balances, ok := farmingPoolBalances[farmingPool]
+		if !ok {
+			balances = k.bankKeeper.GetAllBalances(ctx, farmingPoolAcc)
+			farmingPoolBalances[farmingPool] = balances
+		}
+
+		dc, ok := distrCoins[farmingPool]
+		if !ok {
+			dc = make(map[uint64]sdk.Coins)
+			distrCoins[farmingPool] = dc
+		}
+
+		switch plan := plan.(type) {
+		case *types.FixedAmountPlan:
+			dc[plan.GetId()] = plan.EpochAmount
+		case *types.RatioPlan:
+			dc[plan.GetId()], _ = sdk.NewDecCoinsFromCoins(balances...).MulDecTruncate(plan.EpochRatio).TruncateDecimal()
+		}
+	}
+
+	var distrInfos []DistributionInfo
+	for farmingPool, coins := range distrCoins {
+		totalCoins := sdk.NewCoins()
+		for _, amt := range coins {
+			totalCoins = totalCoins.Add(amt...)
+		}
+
+		balances := farmingPoolBalances[farmingPool]
+		fmt.Println("farmingPool=", farmingPool, "balances=", balances, "totalCoins=", totalCoins)
+		if !totalCoins.IsAllLT(balances) {
+			continue
+		}
+
+		for planID, amt := range coins {
+			distrInfos = append(distrInfos, DistributionInfo{
+				Plan:   plans[planID],
+				Amount: amt,
+			})
+		}
+	}
+
+	return distrInfos
+}
+
+func (k Keeper) DistributeRewards(ctx sdk.Context) error {
+	stakingsByDenom := make(map[string][]types.Staking)
+	totalStakedAmtByDenom := make(map[string]sdk.Int)
+
+	for _, distrInfo := range k.DistributionInfos(ctx) {
+		stakingCoinWeights := distrInfo.Plan.GetStakingCoinWeights()
+		totalDistrAmt := sdk.NewCoins()
+
+		totalWeight := sdk.ZeroDec()
+		for _, coinWeight := range stakingCoinWeights {
+			totalWeight = totalWeight.Add(coinWeight.Amount)
+		}
+
+		for _, coinWeight := range stakingCoinWeights {
+			stakings, ok := stakingsByDenom[coinWeight.Denom]
+			if !ok {
+				stakings = k.GetStakingsByStakingCoinDenom(ctx, coinWeight.Denom)
+				stakingsByDenom[coinWeight.Denom] = stakings
+
+				for _, staking := range stakings {
+					totalStakedAmtByDenom[coinWeight.Denom] = totalStakedAmtByDenom[coinWeight.Denom].Add(staking.StakedCoins.AmountOf(coinWeight.Denom))
+				}
+			}
+
+			totalStakedAmt := totalStakedAmtByDenom[coinWeight.Denom]
+
+			for _, staking := range stakings {
+				stakedAmt := staking.StakedCoins.AmountOf(coinWeight.Denom)
+				if !stakedAmt.IsPositive() {
+					continue
+				}
+
+				stakedProportion := stakedAmt.ToDec().QuoTruncate(totalStakedAmt.ToDec())
+				weightProportion := coinWeight.Amount.QuoTruncate(totalWeight)
+				distrAmt, _ := sdk.NewDecCoinsFromCoins(distrInfo.Amount...).MulDecTruncate(stakedProportion.MulTruncate(weightProportion)).TruncateDecimal()
+
+				k.SetReward(ctx, coinWeight.Denom, staking.GetFarmer(), distrAmt)
+				totalDistrAmt = totalDistrAmt.Add(distrAmt...)
+			}
+		}
+
+		if err := k.bankKeeper.SendCoins(ctx, distrInfo.Plan.GetFarmingPoolAddress(), distrInfo.Plan.GetRewardPoolAddress(), totalDistrAmt); err != nil {
```

Comment:
Can we set `LastDistributedTime` and `TotalDistributedRewardCoins` values for well-distributed plans? 
To do so, Key functions will need to be added keys.go and plans.go with Get, Set functions for these.

Reply:
```
- LastDistributedTime: `0x13 | Id -> time.Time`
- TotalDistributedRewardCoins: `0x14 | PlanId | StakingCoinDenomLen (1 byte) | StakingCoinDenom → ProtocolBuffer(sdk.Coins)`
```

Reply:
Ok, I'll add those.

Reply:
I'll work on this on a separate issue.